examples/ex18.cpp

```diff
@@ -0,0 +1,304 @@
+   ODESolver *ode_solver = NULL;
+   switch (ode_solver_type)
+   {
+      case 1: ode_solver = new ForwardEulerSolver; break;
+      case 2: ode_solver = new RK2Solver(1.0); break;
+      case 3: ode_solver = new RK3SSPSolver; break;
+      case 4: ode_solver = new RK4Solver; break;
+      case 6: ode_solver = new RK6Solver; break;
+      default:
+         cout << "Unknown ODE solver type: " << ode_solver_type << '\n';
+         return 3;
+   }
+
+   // 4. Refine the mesh to increase the resolution. In this example we do
+   //    'ref_levels' of uniform refinement, where 'ref_levels' is a
+   //    command-line parameter.
+   for (int lev = 0; lev < ref_levels; lev++)
+   {
+      mesh.UniformRefinement();
+   }
+
+   // 5. Define the discontinuous DG finite element space of the given
+   //    polynomial order on the refined mesh.
+   DG_FECollection fec(order, dim);
+   // Finite element space for a scalar (thermodynamic quantity)
+   FiniteElementSpace fes(&mesh, &fec);
+   // Finite element space for a mesh-dim vector quantity (momentum)
+   FiniteElementSpace dfes(&mesh, &fec, dim, Ordering::byNODES);
+   // Finite element space for all variables together (total thermodynamic state)
+   FiniteElementSpace vfes(&mesh, &fec, num_equation, Ordering::byNODES);
+
+   // This example depends on this ordering of the space.
+   MFEM_ASSERT(fes.GetOrdering() == Ordering::byNODES, "");
+
+   cout << "Number of unknowns: " << vfes.GetVSize() << endl;
+
+   // 6. Define the initial conditions, save the corresponding mesh
+   //    and grid functions to a file. Note again that the file can be
+   //    opened with GLvis with the -gc option.
+   Array<int> offsets(num_equation + 1);
+   for (int k = 0; k <= num_equation; k++) { offsets[k] = k * vfes.GetNDofs(); }
+   BlockVector u_block(offsets);
+
+   // Density grid function for visualization.
+   GridFunction mom(&dfes, u_block.GetData() + offsets[1]);
+
+   // Initialize the block vector state: define a vector finite element space
+   // for all equations and initialize together.
+   VectorFunctionCoefficient u0(num_equation, u0_function);
+   GridFunction sol(&vfes, u_block.GetData());
+   sol.ProjectCoefficient(u0);
+
+   // Output the initial solution.
+   {
+      ofstream mesh_ofs("vortex.mesh");
+      mesh_ofs.precision(precision);
+      mesh_ofs << mesh;
+
+      for (int k = 0; k < num_equation; k++)
+      {
+         GridFunction uk(&fes, u_block.GetBlock(k));
+         ostringstream sol_name;
+         sol_name << "vortex-" << k << "-init.gf";
+         ofstream sol_ofs(sol_name.str().c_str());
+         sol_ofs.precision(precision);
+         sol_ofs << uk;
+      }
+   }
+
+   // 7. Set up the nonlinear form corresponding to the DG discretization of the
+   //    flux divergence, and assemble the corresponding mass matrix.
+   MixedBilinearForm Aflux(&dfes, &fes);
+   Aflux.AddDomainIntegrator(new DomainIntegrator(dim));
+   Aflux.Assemble();
+
+   NonlinearForm A(&vfes);
```

Comment:
Once I merge my block nonlinear form branch, we can update this to use a true block finite element space.